### PR TITLE
feat(api): stubbing in referral counter

### PR DESCRIPTION
**Pull request by andrewxhill**
adding counter to referral codes to start issuing them to users

### CODE REVIEW

File: central/controllers/referrals.go [hunk 1]
```diff
@@ -1,15 +1,16 @@
...
 	"math/rand"
 	"net/http"
 	"os"
 	"strconv"
 	"time"
+
+	"github.com/gin-gonic/gin"
+	"github.com/globalsign/mgo/bson"
+	"github.com/textileio/textile-go/central/dao"
+	"github.com/textileio/textile-go/central/models"
```
Comment (andrewxhill) [inline central/controllers/referrals.go]: not sure why the imports moved... 🤷‍♂️

File: central/controllers/referrals.go [hunk 2]
```diff
@@ -21,7 +22,15 @@ func CreateReferral(c *gin.Context) {
 
 	// how many should we make?
 	count := 1
+	// how many times should we be able to use them
+	limit := 1
```
Comment (andrewxhill) [inline central/controllers/referrals.go]: takes a `limit` param in the api, defaults to 1 which will be incremented down to zero when the code is used

File: central/dao/dao.go [hunk 3]
```diff
@@ -109,7 +110,7 @@ func (m *DAO) FindReferralByCode(code string) (models.Referral, error) {
 // List referrals
 func (m *DAO) ListUnusedReferrals() ([]models.Referral, error) {
 	var refs []models.Referral
-	err := db.C(referralCollection).Find(bson.M{"used": bson.M{"$eq": nil}}).All(&refs)
+	err := db.C(referralCollection).Find(bson.M{"remaining": bson.M{"$gt": 0}}).All(&refs)
```
Comment (andrewxhill) [inline central/dao/dao.go]: unused would now just be any with Remaining count greater than 0

File: central/models/user.go [hunk 4]
```diff
@@ -40,9 +41,8 @@ type Identity struct {
...
 type Referral struct {
-	ID      bson.ObjectId  `bson:"_id" json:"id"`
-	Code    string         `bson:"code" json:"code"`
-	Created time.Time      `bson:"created" json:"created"`
-	Used    *time.Time     `bson:"used" json:"used"`
-	UserID  *bson.ObjectId `bson:"user_id" json:"user_id"`
+	ID        bson.ObjectId `bson:"_id" json:"id"`
+	Code      string        `bson:"code" json:"code"`
+	Created   time.Time     `bson:"created" json:"created"`
+	Remaining int           `bson:"remaining" json:"remaining"`
```
Comment (andrewxhill) [inline central/models/user.go]: no longer a Used or UserID param on the referrals since they will be used by multiple users

**Comment by sanderpick** [comment]
looking good dude

File: central/controllers/referrals.go (hunk 1 above)
Comment (sanderpick) [reply central/controllers/referrals.go]: for this, if you remove the whitespace in the import blocks, then run `npm run format` it should brings it back

**Comment by andrewxhill** [comment]
okay, so was thinking of adding the referral creation to the user signup step, but i'm tempted instead to just leave it as an api and then throw it into react native to grab a new one the first time it's needed.

sound okay?

**Comment by andrewxhill** [comment]
Okay, this is gtg

**Comment by sanderpick** [comment]
hell yeah. since we don't have ci setup for this yet, i'll have to manually build the docker-compose stack, push that to docker hub, and update the swarm... can do in a bit

**Pull request merged**